json-schema/json-schema/src/test/scala/endpoints4s/TuplerTests.scala

```diff
@@ -5,18 +5,45 @@ class TuplerTests {
   def tupling[A, B](a: A, b: B)(implicit tupler: Tupler[A, B]): tupler.Out =
     tupler(a, b)
 
-  def forall[A, B, C, D, E, F](a: A, b: B, c: C, d: D, e: E, f: F): Unit = {
+  def forall[A, B, C, D, E, F, G, H, I](a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I): Unit = {
+    tupling((), a): A
+    tupling(a, ()): A
     tupling(a, b): (A, B)
-    tupling(a, (b, c)): (A, B, C)
     tupling((a, b), c): (A, B, C)
+    tupling(a, (b, c)): (A, B, C)
+    tupling((a, b, c), d)  : (A, B, C, D)
     tupling((a, b), (c, d)): (A, B, C, D)
-    tupling(a, (b, c, d, e)): (A, B, C, D, E)
+    tupling(a, (b, c, d))  : (A, (B, C, D))
+    tupling((a, b, c, d), e)  : (A, B, C, D, E)
+    tupling((a, b, c), (d, e)): (A, B, C, (D, E))
     tupling((a, b), (c, d, e)): (A, B, C, D, E)
-    tupling((a, b, c), d): (A, B, C, D)
-    tupling((a, b, c, d), e): (A, B, C, D, E)
-    tupling((a, b, c, d, e), f): (A, B, C, D, E, F)
-    tupling(a, ()): A
-    tupling((), a): A
+    tupling(a, (b, c, d, e))  : (A, (B, C, D, E))
+    tupling((a, b, c, d, e), f)  : (A, B, C, D, E, F)
+    tupling((a, b, c, d), (e, f)): (A, B, C, D, (E, F))
```

Comment:
Here as well, I would like to return a flat tuple `(A, B, C, D, E, F)`

Reply:
This would be a change to tupler. I've updated the PR with a reduced diff in favor of easy reviewing, I'll re-sort towards the end of this process.

If at all possible, it would be ideal to permit rebasing and force-push for branches, but if you'll squash at the end I'll just not be concerned about producing unnecessary intermediate commits here.

Reply:
Also, sorry, to what end are we doing this?

You asked about
```scala
    tupling((a, b, c, d), (e, f)): (A, B, C, D, (E, F))
```
but
```scala
    tupling((a, b, c), (d, e, f)): (A, B, C, (D, E, F))
    tupling((a, b), (c, d, e, f)): ((A, B), C, D, E, F)
    tupling(a, (b, c, d, e, f)): (A, (B, C, D, E, F))
```
would also not be covered by this. Isn't this just that we only have so many tupler extensions so far? [I've worked around this when attempting to generate tuple structures compatible with the current Tuple semantics](https://github.com/twilio/guardrail/blob/v0.62.0/modules/codegen/src/main/scala/com/twilio/guardrail/generators/Scala/EndpointsClientGenerator.scala#L290-L294), so I'd err on the side of stability until we have the truly generic solution, if I'm being selfish.

The change in order to make this test pass is
```diff
diff --git a/json-schema/json-schema/src/main/scala/endpoints4s/Tupler.scala b/json-schema/json-schema/src/main/scala/endpoints4s/Tupler.scala
index 929d19e8..88bfcfb2 100644
--- a/json-schema/json-schema/src/main/scala/endpoints4s/Tupler.scala
+++ b/json-schema/json-schema/src/main/scala/endpoints4s/Tupler.scala
@@ -114,7 +114,23 @@ trait Tupler3 extends Tupler2 {
 
 }
 
-trait TuplerLeftUnit extends TuplerAppend {
+trait Tupler4 extends TuplerAppend {
+
+  implicit def tupler4And2[A, B, C, D, E, F]
+      : Tupler[(A, B, C, D), (E, F)] { type Out = (A, B, C, D, E, F) } =
+    new Tupler[(A, B, C, D), (E, F)] {
+      type Out = (A, B, C, D, E, F)
+      def apply(abcd: (A, B, C, D), ef: (E, F)): (A, B, C, D, E, F) =
+        (abcd._1, abcd._2, abcd._3, abcd._4, ef._1, ef._2)
+      def unapply(out: (A, B, C, D, E, F)): ((A, B, C, D), (E, F)) = {
+        val (a, b, c, d, e, f) = out
+        ((a, b, c, d), (e, f))
+      }
+    }
+
+}
+
+trait TuplerLeftUnit extends Tupler4 {
 
   implicit def leftUnit[A]: Aux[Unit, A, A] =
     new Tupler[Unit, A] {
diff --git a/json-schema/json-schema/src/test/scala/endpoints4s/TuplerTests.scala b/json-schema/json-schema/src/test/scala/endpoints4s/TuplerTests.scala
index 67c00c4d..4f630166 100644
--- a/json-schema/json-schema/src/test/scala/endpoints4s/TuplerTests.scala
+++ b/json-schema/json-schema/src/test/scala/endpoints4s/TuplerTests.scala
@@ -20,7 +20,7 @@ class TuplerTests {
 
     tupling(a, (b, c, d)): (A, (B, C, D))
     tupling((a, b, c), (d, e)): (A, B, C, (D, E))
-    tupling((a, b, c, d), (e, f)): (A, B, C, D, (E, F))
+    tupling((a, b, c, d), (e, f)): (A, B, C, D, E, F)
     tupling((a, b, c), (d, e, f)): (A, B, C, (D, E, F))
     tupling((a, b), (c, d, e, f)): ((A, B), C, D, E, F)
     tupling(a, (b, c, d, e, f)): (A, (B, C, D, E, F))
```
but adding the various other combinations, `1And5`, `3And3`, `4And2`, made a bunch of other tests fail so I did not include it here.

Reply:
> Also, sorry, to what end are we doing this?
> 
> You asked about
> 
> ```scala
>     tupling((a, b, c, d), (e, f)): (A, B, C, D, (E, F))
> ```
> 
> but
> 
> ```scala
>     tupling((a, b, c), (d, e, f)): (A, B, C, (D, E, F))
>     tupling((a, b), (c, d, e, f)): ((A, B), C, D, E, F)
>     tupling(a, (b, c, d, e, f)): (A, (B, C, D, E, F))
> ```
> 
> would also not be covered by this. Isn't this just that we only have so many tupler extensions so far?

The goal of `Tupler` is to concatenate tuples, so that we end-up with a flat tuple instead of nested tuples (and occurrences of `Unit` are removed). The cases you mention should be fixed as well… The problem is that the `Tupler` instances currently cover only the most common cases, and being exhaustive would require a lot of implicit definitions (and headaches).

I suggest that we remove these lines from the test, as this is not something we want to support (I believe).

Reply:
Maybe split them out into a "stability" section? The reason for this PR in the first place is because an otherwise helpful addition ended up making some valid usecases completely unrepresentable.

At some point this breaks down, for instance a `Tuple22` combined with anything. I've needed to work around this for particularly complex OpenAPI specifications, which is what led me to `endpoints` in the first place.

Either way, your points about defining disambiguating `Tupler` instances in my own project are completely valid, just looking to wrap up this issue to help out others that may run into this same problem before the HList matter is settled.

I'll move the tests to a separate section that can act as a canary for when things change, but if you feel they should be removed I'm perfectly happy to do that as well.

Thanks again for your time here.